### ISSUE TITLE
New Java8 Date class: JsonMappingException: No Record or Map type found in union type: ["null","string"]

### ISSUE THREAD

**Issue by asmaier** [FasterXML/jackson-dataformat-avro#38]
It seems to me that the new Java8 Date classes are not handled correctly. When I try to generate a schema from my File class below and use that schema to write an Avro message, I get an JsonMappingException: No Record or Map type found in union type: ["null","string"]:

```
    private static class File {
        public String filename = "TestFile.txt";
        public OffsetDateTime dateTime = OffsetDateTime.now();
    }

@Test
    public void testAvroSchemaGenerationWithJackson() throws JsonProcessingException {

        ObjectMapper mapper = new ObjectMapper(new AvroFactory());
        mapper.registerModule(new JavaTimeModule());
        mapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);

        AvroSchemaGenerator visitor = new AvroSchemaGenerator();
        mapper.acceptJsonFormatVisitor(File.class, visitor);
        AvroSchema schema = visitor.getGeneratedSchema();

        ObjectMapper mapper2 = new ObjectMapper(new AvroFactory());
        System.out.println(mapper2.writer(schema).writeValueAsBytes(new File()));
    }
```

The generated Avro schema looks like this
```
{
  "type" : "record",
  "name" : "File",
  "namespace" : "xx.xxx.xxx",
  "doc" : "Schema for xx.xxx.xxx.SchemaGenerationTest$File",
  "fields" : [ {
    "name" : "dateTime",
    "type" : [ "null", "string" ]
  }, {
    "name" : "filename",
    "type" : [ "null", "string" ]
  } ]
}
```

**Comment by cowtowncoder** [comment]
Hmmh. Now this is interesting. For some reason `OffsetDateTime` is trying to serialize itself as an Object (which you can see if using plain vanilla JSON `ObjectMapper`). This is what causes discrepancy; schema indicates that a `String` is expected. Error message is bit misleading as well, I'll see if I can improve it.

But more interesting question is why proper serializer is not found...

**Comment by cowtowncoder** [comment]
Actually, never mind; my test wrt JSON was without registering serializer. So there's something else going on there.

**Comment by cowtowncoder** [comment]
Ah-ha. So the problem wrt serialization as Object is because test case uses different `ObjectMapper`, one that is NOT configured to handle `OffsetDateTime`: `mapper2` vs `mapper`.

Registering `JavaTimeModule` helps a bit, although does expose different problem.

**Comment by cowtowncoder** [comment]
Actually a fix in `jsr310` datatype module in 2.7.3 should resolve this.
